KOGITO-2303: Use vscode-extension-tester to test the VS Code Extension

domhanak opened the pull request:
TODO: General refactor to multiple files, fix some dependencies I added by mistake, add scesim smoke test

Ready for initial feedback and comments, I am working on refactoring this in the meantime so any comments will be incorporated to into it.

domhanak commented:
Tests are failing now as new version of VSCode was released and the editors are not loading properly.

caponetto commented on packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/resources/demo.dmn:
```diff
@@ -0,0 +1 @@
+
```
Is it supposed to be empty?

caponetto commented on packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/extension-marketplace.test.ts:
```diff
@@ -0,0 +1,126 @@
+import { EditorView, WebView, ExtensionsViewItem, By, WebElement } from 'vscode-extension-tester';
```
Please add the copyright header to all source files.

caponetto commented on packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/extension-marketplace.test.ts:
```diff
@@ -0,0 +1,126 @@
+
+		it('Find extension', async function () {
+			this.timeout(10000);
+			item = await marketplace.findExtension(`@installed ${pjson.displayName}`) as ExtensionsViewItem;
```
Shouldn't this code run inside `beforeAll`?

domhanak replied:
This test that we can find the extension in marketplace and it installs without problems. I can move it, but the idea is to check general naming & presence & apperance on marketplace

caponetto commented on packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/extension-marketplace.test.ts:
```diff
@@ -0,0 +1,126 @@
+		let item: ExtensionsViewItem;
+
+		before(async function () {
+			this.timeout(10000);
```
Where does `this.timeout` come from? I guess I'm missing something here.

domhanak replied:
This wiil be updated

caponetto commented on packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/extension-marketplace.test.ts:
```diff
@@ -0,0 +1,126 @@
+		let marketplace: Marketplace;
+		let item: ExtensionsViewItem;
+
+		before(async function () {
```
Consider using an arrow function here (and in the other applicable places below) just like you did on line 24.

domhanak replied:
I would like to have a control of timeout of each test case, the operations may wary and some tasks need more time than others.

caponetto commented on packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/extension-marketplace.test.ts:
```diff
@@ -0,0 +1,126 @@
+
+		after(async () => {
+			await marketplace.close();
+			await new EditorView().closeAllEditors();
```
Can we have only one instance of EditorView? It is being created five times on this file.

caponetto commented on packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/extension-marketplace.test.ts:
```diff
@@ -0,0 +1,126 @@
+		})
+
+		it('Open demo.bpmn file in BPMN Editor', async () => {
+			const kogitoFrame: WebElement = await webview.findWebElement(By.id('org.kie.workbench.common.stunner.kogito.KogitoBPMNEditor'))
+			assert.isDefined(kogitoFrame);
+			const palette: WebElement =  await webview.findWebElement(By.className("kie-palette"));
+			const properties: WebElement = await webview.findWebElement(By.className("qe-docks-item-E-DiagramEditorPropertiesScreen"));
+			const explorer: WebElement = await webview.findWebElement(By.className("qe-docks-item-E-ProjectDiagramExplorerScreen"))
+			assert.isTrue(await palette.isDisplayed() && await palette.isEnabled());
+			assert.isTrue(await properties.isDisplayed() && await properties.isEnabled());
+			assert.isTrue(await explorer.isDisplayed() && await explorer.isEnabled());
+
+			const exploreDiagramButton = await explorer.findElement(By.xpath('//button[@data-title=\'Explore Diagram\']'));
+			assert.isTrue(await exploreDiagramButton.isDisplayed() && await exploreDiagramButton.isEnabled())
+			await exploreDiagramButton.click();
+			assert.isTrue(await explorer.findElement(By.xpath(aComponentWithText('demo'))).isDisplayed());
+			assert.isTrue(await explorer.findElement(By.xpath(aComponentWithText('Start'))).isDisplayed());
+			assert.isTrue(await explorer.findElement(By.xpath(aComponentWithText('End'))).isDisplayed());			
```
Consider rearranging the code of this test using the AAA pattern. IMO it'd improve the organization.

caponetto commented on packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/extension-marketplace.test.ts:
```diff
@@ -0,0 +1,126 @@
+
+		let webview : WebView;
+
+		before(async function () {
+			this.timeout(25000)
+			await Dialog.openFile(path.join(RESOURCES, DEMO_DMN));
+			webview = new WebView(new EditorView(), By.linkText(DEMO_DMN));
+			await webview.switchToFrame();
+		});
+
+		after(async () => {
+			await webview.switchBack()
+			await new EditorView().closeAllEditors();
+		})
```
The code inside `before` and `after` is almost identical to the previous test `BPMN Editor is loading`.
So consider creating reusable methods to avoid code duplication.

caponetto commented on packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/extension-marketplace.test.ts:
```diff
@@ -0,0 +1,126 @@
+		});
+	});
+
+	describe("BPMM Editor is loading", () => {
```
```suggestion
	describe("BPMN Editor is loading", () => {
```

caponetto commented on packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/extension-marketplace.test.ts:
```diff
@@ -0,0 +1,126 @@
+			assert.isTrue(await exploreDiagramButton.isDisplayed() && await exploreDiagramButton.isEnabled());
+		});
+	})
+});
```
Are we gonna have similar tests for `scesim` on this PR?

domhanak replied:
@caponetto Yes,  I ll add them, there wasn't scesim when I started this.
I ll ping you when ready! :)

caponetto reviewed (commented):
Hi @domhanak, thanks for the PR!
I've left some comments, please check if they make sense.
Also, there are some conflicts that need to be resolved before merging this PR.

+1 for Tiago's comments.

domhanak commented:
Couple of small refactors and improvements is missing, but overall it is ready for the next round of review. Please let me know @caponetto @paulovmr 

tsconfig.json updates will be reverted.

tiagobento commented on package.json:
```diff
@@ -27,6 +27,8 @@
     "@testing-library/react": "^10.0.2",
     "@testing-library/react-hooks": "^3.2.1",
+    "@types/chai": "^4.2.11",
+    "@types/enzyme": "3.10.0",
```
Do we need enzyme?

domhanak replied:
i double check, thanks

tiagobento commented on packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/helpers/VSCodeTestHelper.ts:
```diff
@@ -0,0 +1,81 @@
+        // const input = await InputBox.create();
+        // await input.selectQuickPick('KIE Kogito Editors');
+        const webview = new WebView(new EditorView(), By.linkText(fileName));
+        await DefaultWait.sleep(10000);
```
This seems to be the only usage we have of `vscode-uitests-tooling`. Looks to me that it's just a Promise being resolved after a setTimeout executes.. can we refactor that to remove the dependency with `vscode-uitests-tooling`?

domhanak replied:
Well we will definitely use more functionalities from that package, the package is a collection of extra functionalities on top of vscode-extension-tester.

tiagobento reviewed (changes requested):
Thanks @domhanak! This is looking great! Please consider my comments inline. After you revert the changes on tsconfig.json, we should be good to go! Nice work!

domhanak commented:
@tiagobento Ready now, I kept two settings in tsconfig of vscode's package.
`--resolveJsonModule` to allow imports of json file, very usefull when we want to introduce tests to verify that the extension name is same in marketplace as in package.json
`outDir` to make sure that the outputed JS files are in one place and dont pollute the workspace

This is now ready for merging if you approve, I have couple of new tests cases in progress along with some code improvements and refactors coming after this.
For example, save operation, service tasks, backend notification tests.

caponetto commented:
Hi @domhanak, can you please add the copyright header to the source files?

domhanak commented:
@caponetto Done, how do I rerun the builds?

caponetto commented:
Thanks, @domhanak! I just re-run them.

tiagobento commented:
@domhanak Regular CI is failing, can you please check that?

domhanak commented:
@tiagobento I checked, seems like an instability. I couldn't reproduce on subsequent runs locally, and next build succeeded.
I rebased on master and squashed the commits. Let me know if we can merge.

domhanak commented:
FDB failed on an issue with integration tests, not related to this PR.

caponetto reviewed (approved):
LGTM

tiagobento commented:
@tomasdavidorg That's a different error from the ones we've seen before. Do you mind checking it out?

tomasdavidorg commented:
> @tomasdavidorg That's a different error from the ones we've seen before. Do you mind checking it out?

I will take a look. Looks like the driver was not created / Chrome was not initialized. It is not related to this PR.

jomarko commented on packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/resources/incident-response.bpmn2:
```diff
@@ -0,0 +1,1502 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
is this file used in some test?

jomarko reviewed (commented):
Tests looks good to me. Just few questions. Good job @domhanak

tomasdavidorg commented on packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/helpers/CommonAsserts.ts:
```diff
@@ -0,0 +1,27 @@
+ * @param tested WebElement to test
+ */
+export const assertWebElementIsDisplayedEnabled = async (tested: WebElement): Promise<void> => {
+    assert.isTrue(await tested.isDisplayed() && await tested.isEnabled());
```
That should be correct. When you have `async` function returning void then the return type is `Promise<void>`.

jomarko replied:
ok, thank you

domhanak commented on packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/extension-editors.test.ts:
```diff
@@ -0,0 +1,103 @@
+    let folderView : SideBarView;
+
+    before(async function() {
+        this.timeout(60000);
```
This is not dynamic waiting, it just sets the timeout for this function to 60s. If the `before()` setup does not complete in 60s the test fails.

tomasdavidorg replied:
:man_facepalming: Sorry. :smile:

domhanak commented on packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/helpers/BpmnEditorTestHelper.ts:
```diff
@@ -0,0 +1,111 @@
+     * 
+     * Verifies the button is displayed and enabled.
+     * Verifies there is an expanded panel element displayed and enabled.
+     * Verifies there is a <h3> element with propert text.
```
What do you mean `a <h3> element`?

tomasdavidorg replied:
I think @jomarko had in mind `propert` to `proper`.

domhanak commented on packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/extension-editors.test.ts:
```diff
@@ -0,0 +1,103 @@
+import ScesimEditorTestHelper from './helpers/ScesimEditorTestHelper';
+import DmnEditorTestHelper from './helpers/DmnEditorTestHelper';
+
+describe("Editors are loading properly", () => {
```
1. binds `this` to the method itself not to the context. Used  to control timeout per method/test. For example in `before` `after`.
Not sure what is the go-to option. There a pros/cons for both approaches so I would probably let it be this time and do a meeting to discuss this.

domhanak commented:
Build failed on mac that couldn't download the dependencies... @caponetto How do I re-run the jobs myself? Is there something like `jenkins retest this` ?

caponetto commented:
@domhanak There's a button to re-run the jobs but I think only available for users with write permission in the repository.
I'll re-run for you but one workaround for re-running it whenever you want would be force pushing a new commit hash, like:
`git commit --amend --no-edit && git push -f <remove> <branch>`

domhanak commented:
@caponetto that is not very user-friendly, but I wil ltry it. The approach could be made less strict so that at least the creator of PR could re trigger the build. Anyway, `macos-lastest` failed again with same issue, fail during dependencies download.

tiagobento commented:
@domhanak `macos-latest` failing to download the dependencies sometimes is a known issue. There's no workaround for that at the moment :/ Also, GitHub has not yet made it possible for the owners of the PR to re-run a workflow. When GH Actions was launched, re-running was not possible even for admins, so we also had to use this trick of pushing an empty commit :)

domhanak commented on packages/vscode-extension-pack-kogito-kie-editors/src/ui-test/extension-editors.test.ts:
```diff
@@ -0,0 +1,103 @@
+describe("Editors are loading properly", () => {
+
+    const RESOURCES: string = path.resolve('src', 'ui-test', 'resources');
+	const DEMO_BPMN: string = 'demo.bpmn';
```
Not sure why but I see this as fixed locally :confused:

domhanak commented:
green, let's merge if all good. @tiagobento
